jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/FilePropertiesStack.java

```diff
@@ -0,0 +1,114 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.cli.buildfile;
+
+import com.google.cloud.tools.jib.api.buildplan.FilePermissions;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A class that keeps track of permissions for various stacking file permissions settings in {@link
+ * LayerSpec}.
+ */
+@VisibleForTesting
+class FilePropertiesStack {
+
+  // TODO perhaps use a fixed size list here
+  private List<FilePropertiesSpec> stack = new ArrayList<>(3);
+
+  @Nullable private FilePermissions filePermissions;
+  @Nullable private FilePermissions directoryPermissions;
+  @Nullable private Instant modificationTime;
+  @Nullable private String ownership;
+
+  // intermediate values
+  @Nullable private String user;
+  @Nullable private String group;
+
+  /**
+   * Add a new layer to the file properties stack. When adding a new layer, it is given highest
+   * priority when resolving properties. All values are recalculated.
+   */
+  public void push(FilePropertiesSpec filePropertiesSpec) {
+    Preconditions.checkState(
+        stack.size() < 3, "Error in file properties stack push, stacking over 3");
+    stack.add(filePropertiesSpec);
+    updateProperties();
+  }
+
+  /** Remove the last layer from the stack. All values are recalculated. */
+  public void pop() {
+    Preconditions.checkState(stack.size() > 0, "Error in file properties stack pop, popping at 0");
+    stack.remove(stack.size() - 1);
+    updateProperties();
+  }
+
+  private void updateProperties() {
+    // clear existing permissions before recalculating
+    filePermissions = null;
+    directoryPermissions = null;
+    modificationTime = null;
+    ownership = null;
+    user = null;
+    group = null;
+
+    // the item with the lowest index has the lowest priority
+    for (FilePropertiesSpec properties : stack) {
+      properties.getFilePermissions().ifPresent(permissions -> filePermissions = permissions);
+      properties
+          .getDirectoryPermissions()
+          .ifPresent(permissions -> directoryPermissions = permissions);
+      properties.getTimestamp().ifPresent(timestamp -> modificationTime = timestamp);
+      properties.getUser().ifPresent(user -> this.user = user);
+      properties.getGroup().ifPresent(group -> this.group = group);
+    }
+    // ownership calculations
+    if (group == null) {
```

Comment:
So if a lower stack had `user=1234` but a higher stack `group=5678`, it will lost `1234`. Probably this is what we intend, but I just wanted to confirm. Maybe good to have a test to formalize the intention.

Reply:
I'm not sure I understand, the ownership calculation is outside the loop. So once `group` and `user` are finalized, then the values are converted into an ownership string.

Reply:
Ah, I misunderstood. Yeah, the calculation is outside the loop.

Just so you know, jib-core will probably accept `:` (a single colon) as an ownership string, but I think it makes sense to return null as done here when both `user` and `group` are null.